feedback/ngffeedback.cpp

```diff
@@ -0,0 +1,326 @@
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301 USA
+ */
+
+#include <qfeedbackeffect.h>
+#include <QCoreApplication>
+#include <QtPlugin>
+#include "ngffeedback.h"
+
+Q_LOGGING_CATEGORY(ngflc, "qt.Feedback.ngf", QtWarningMsg)
+
+NGFFeedback::NGFFeedback(QObject *parent)
+    : QObject(parent)
+    , QFeedbackHapticsInterface()
+    , QFeedbackThemeInterface()
+    , m_actuator(createFeedbackActuator(this, 2))
+    , m_actuatorEnabled(true)
+    , m_client(this)
+{
+    qCDebug(ngflc) << "Initializing plugin";
+
+    if (!m_client.connect()) {
+        qCCritical(ngflc) << "Unable to connect to NGFD";
+    }
+
+    connect(&m_client, &Ngf::Client::eventFailed,
+            this, &NGFFeedback::failed);
+    connect(&m_client, &Ngf::Client::eventCompleted,
+            this, &NGFFeedback::completed);
+    connect(&m_client, &Ngf::Client::eventPlaying,
+            this, &NGFFeedback::playing);
+    connect(&m_client, &Ngf::Client::eventPaused,
+            this, &NGFFeedback::paused);
+
+    m_effects[QFeedbackEffect::Press] = QStringLiteral("feedback_press");
+    m_effects[QFeedbackEffect::Release] = QStringLiteral("feedback_release");
+    m_effects[QFeedbackEffect::PressWeak] = QStringLiteral("feedback_press_weak");
+    m_effects[QFeedbackEffect::ReleaseWeak] = QStringLiteral("feedback_release_weak");
+    m_effects[QFeedbackEffect::PressStrong] = QStringLiteral("feedback_press_strong");
+    m_effects[QFeedbackEffect::ReleaseStrong] = QStringLiteral("feedback_release_strong");
+    m_effects[QFeedbackEffect::DragStart] = QStringLiteral("feedback_drag_start");
+    m_effects[QFeedbackEffect::DragDropInZone] = QStringLiteral("feedback_drag_drop_in_zone");
+    m_effects[QFeedbackEffect::DragDropOutOfZone] = QStringLiteral("feedback_drag_drop_out_of_zone");
+    m_effects[QFeedbackEffect::DragCrossBoundary] = QStringLiteral("feedback_drag_cross_boundary");
+    m_effects[QFeedbackEffect::Appear] = QString();
+    m_effects[QFeedbackEffect::Disappear] = QString();
+    m_effects[QFeedbackEffect::Move] = QString();
+}
+
+NGFFeedback::~NGFFeedback()
+{
+    qCDebug(ngflc) << "Deinitializing plugin";
+}
+
+void NGFFeedback::failed(quint32 id)
+{
+    ActiveEffect *active = findEffect(id);
+    if (active) {
+        qCWarning(ngflc) << "Effect failed, id" << id;
+        m_activeEffects.removeAll(*active);
+    }
+}
+
+void NGFFeedback::completed(quint32 id)
+{
+    ActiveEffect *active = findEffect(id);
+    if (active) {
+        qCDebug(ngflc) << "Effect completed, id" << id;
+        m_activeEffects.removeAll(*active);
+    }
+}
+
+void NGFFeedback::playing(quint32 id)
+{
+    ActiveEffect *active = findEffect(id);
+    if (active) {
+        qCDebug(ngflc) << "Effect playing, id" << id;
+        active->state = QFeedbackEffect::Running;
+    }
+}
+
+void NGFFeedback::paused(quint32 id)
+{
+    ActiveEffect *active = findEffect(id);
+    if (active) {
+        qCDebug(ngflc) << "Effect paused, id" << id;
```

Comment:
I mean these could be outside the if, there shouldn't be a case where invalid id would be logged here.

Reply:
I changed this so that it changes the state and then just logs the id. I had not realized that the ids from ngfd are app specific. It's nice that apps can't touch the effects from each other.